pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -571,27 +581,46 @@ func (p *Provider) makeGatewayStatus(listenerStatuses []v1alpha1.ListenerStatus)
 	return gatewayStatus, nil
 }
 
-func hostRule(httpRouteSpec v1alpha1.HTTPRouteSpec) string {
-	rules := []string{}
-	ruleType := "Host"
+func hostRule(httpRouteSpec v1alpha1.HTTPRouteSpec) (string, error) {
+	var hostNames []string
+	var hostRegexNames []string
+
 	for _, hostname := range httpRouteSpec.Hostnames {
 		host := string(hostname)
 		if host == "" {
 			continue
 		}
-		if strings.HasPrefix(host, "*.") {
-			ruleType = "HostRegexp"
-			host = strings.Replace(host, "*.", "{subdomain:[a-zA-Z0-9-]+}.", 1)
+
+		wildcard := strings.Count(host, "*")
+		if wildcard >= 1 {
+			// https://gateway-api.sigs.k8s.io/spec/#networking.x-k8s.io/v1alpha1.Hostname
+			if !strings.HasPrefix(host, "*.") || wildcard > 1 {
```

Comment:
I'm pretty sure that this will work incorrectly if there are 2 wildcards no? the OR will always pass?

could this be reworked as a fast fail instead of nested ifs?

```
if !strings.HasPrefix(host, "*.") && wildcard > 0 {
// fail if the wildcard is not at the start
return error
}

if wildcard > 1 {
// use hostregex
} else {
//use host
}

Reply:
If the fail on 2 wildcards is intended, then this would do this:

```
if !strings.HasPrefix(host, "*.") && wildcard > 0 {
// fail if the wildcard is not at the start
return error
}

switch wildcard {
case wildcard > 1 :
// fail that there are too many wildcards
break
case wildcard = 1:
// use hostregex
break
default:
//use host
break
}

Reply:
Multiple wildcards are not allowed and the wildcard must be at the front of the domain `*.foo.com`. It basically is the fail early that you mention, I'm not sure there's much practical difference in the change you propose.